File: hard/899/orderly_queue.py

class Solution:
    def orderlyQueue(self, S: str, K: int) -> str:

        N = len(S)

        if K > 1:
            return "".join(sorted(S))

        k = 0
        for j in range(1, N):
            if S[k] > S[j]:
                k = j

        s = S
        c = S[k]
        for j in range(k, N):
            if c == S[k]: 
                S = S[k:] + S[0:k]
                k = 0
                if s > S:
                    s = S
            k += 1

        return s

File: hard/899/orderly_queue.cpp

class Solution {
public:
    string orderlyQueue(string S, int K) {
        int N, k;
        string s;
        char c;
        N = S.size();

        if (K > 1) {
            sort(S.begin(), S.end());
            return S;
        }

        k = 0;
        for (int j = 1; j < N; j++) {
            if (S[k] > S[j]) {
                k = j;
            }
        }

        s = S;
        c = S[k];
        for (int j = k; j < N; j++, k++) {
            if (c == S[k]) {
                S = S.substr(k) + S.substr(0, k);
                k = 0;
                if (lexicographical_compare(S.begin(), S.end(), s.begin(), s.end()) > 0) {
                    s = S;
                }
            }
        }

        return s;
    }
};
